src/modules/power_profiles_daemon.cpp

```diff
@@ -0,0 +1,214 @@
+  Glib::Variant<ProfilesType> profilesVariant;
+  powerProfilesProxy_->get_cached_property(profilesVariant, "Profiles");
+  Glib::ustring name;
+  Glib::ustring driver;
+  Profile profile;
+  for (auto& variantDict : profilesVariant.get()) {
+    if (auto p = variantDict.find("Profile"); p != variantDict.end()) {
+      name = p->second.get();
+    }
+    if (auto d = variantDict.find("Driver"); d != variantDict.end()) {
+      driver = d->second.get();
+    }
+    profile = {name, driver};
+    availableProfiles_.push_back(profile);
+  }
+
+  // Find the index of the current activated mode (to toggle)
+  std::string str = profileStr.get();
+  switchToProfile(str);
+
+  update();
+}
+
+void PowerProfilesDaemon::profileChangedCb(
+    const Gio::DBus::Proxy::MapChangedProperties& changedProperties,
+    const std::vector<Glib::ustring>& invalidatedProperties) {
+  // We're likely connected if this callback gets triggered.
+  // But better be safe than sorry.
+  if (connected_) {
+    if (auto activeProfileVariant = changedProperties.find("ActiveProfile");
+        activeProfileVariant != changedProperties.end()) {
+      std::string activeProfile =
+          Glib::VariantBase::cast_dynamic<Glib::Variant<std::string>>(activeProfileVariant->second)
+              .get();
+      switchToProfile(activeProfile);
+      update();
+    }
+  }
+}
+
+// Look for the profile str in our internal profiles list. Using a
+// vector to store the profiles ain't the smartest move
+// complexity-wise, but it makes toggling between the mode easy. This
+// vector is 3 elements max, we'll be fine :P
+void PowerProfilesDaemon::switchToProfile(std::string const& str) {
+  auto pred = [str](Profile const& p) { return p.name == str; };
+  this->activeProfile_ = std::find_if(availableProfiles_.begin(), availableProfiles_.end(), pred);
+  if (activeProfile_ == availableProfiles_.end()) {
+    throw std::runtime_error("FATAL, can't find the active profile in the available profiles list");
+  }
+}
+
+auto PowerProfilesDaemon::update() -> void {
+  if (connected_) {
+    auto profile = (*activeProfile_);
+    // Set label
+    fmt::dynamic_format_arg_store<fmt::format_context> store;
+    store.push_back(fmt::arg("profile", profile.name));
+    store.push_back(fmt::arg("driver", profile.driver));
+    store.push_back(fmt::arg("icon", getIcon(0, profile.name)));
+    label_.set_markup(fmt::vformat(format_, store));
+    if (tooltipEnabled()) {
+      label_.set_tooltip_text(fmt::vformat(tooltipFormat_, store));
+    }
+
+    // Set CSS class
+    if (!currentStyle_.empty()) {
+      label_.get_style_context()->remove_class(currentStyle_);
+    }
+    label_.get_style_context()->add_class(profile.name);
+    currentStyle_ = profile.name;
+
+    ALabel::update();
+  }
+}
+
+bool PowerProfilesDaemon::handleToggle(GdkEventButton* const& e) {
+  if (connected_) {
+    if (e->type == GdkEventType::GDK_BUTTON_PRESS && powerProfilesProxy_) {
```

Comment:
The check for `powerProfilesProxy_` could be redundant if you already check `connected_`